src/engine/Minus.cpp

```diff
@@ -96,18 +96,24 @@ size_t Minus::getCostEstimate() {
 }
 
 // _____________________________________________________________________________
-auto Minus::makeUndefRangesChecker(bool left) const {
+auto Minus::makeUndefRangesChecker(bool left, const IdTable& idTable) const {
   std::variant<ad_utility::Noop, ad_utility::FindSmallerUndefRanges> findUndef;
   const auto& operation = left ? _left : _right;
-  for (const auto& cols : _matchedColumns) {
-    const auto& [_, info] =
-        operation->getVariableAndInfoByColumnIndex(cols[!left]);
-    // Use expensive operation if one of the columns might contain undef.
-    if (info.mightContainUndef_ !=
-        ColumnIndexAndTypeInfo::UndefStatus::AlwaysDefined) {
-      findUndef.emplace<1>();
-      break;
-    }
+  bool containsUndef = ql::ranges::any_of(
+      _matchedColumns, [&operation, left, &idTable](const auto& cols) {
+        size_t tableColumn = cols[static_cast<size_t>(!left)];
+        const auto& [_, info] =
+            operation->getVariableAndInfoByColumnIndex(tableColumn);
+        if (info.mightContainUndef_ !=
+            ColumnIndexAndTypeInfo::UndefStatus::AlwaysDefined) {
+          return ql::ranges::any_of(idTable.getColumn(tableColumn),
+                                    &Id::isUndefined);
+        }
+        return false;
```

Comment:
What about

```
bool alwaysDefined = mightContain == AlwaysDefined;
return alwaysDefined || ranges::none_of(is undefined)
```
(This of course flips the semantics and your outer call must contain `all_of` and the boll is not `containsUndef` but `alwaysDefined`.

(The nested if with inner + outer returns is hard to reason about).